README
Crypt::Keyczar version 0.07
===========================

Keyczar is an open source cryptographic toolkit designed to make it
easier and safer for devlopers to use cryptography in their
applications. Keyczar supports authentication and encryption with
both symmetric and asymmetric keys. Some features of Keyczar include:

 * A simple API

 * Key rotation and versioning

 * Safe default algorithms, modes, and key lengths

 * Automated generation of initialization vectors and ciphertext
   signatures

 * Perl, Java, Python, and C++ implementations


INSTALLATION
------------
To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

QUICK START
-----------
1. Create new 'keyset' by 'keyczar' tool:

  $ keyczar create --location=/path/to/keyset --purpose=crypt

2. Add new key to 'keyset':

  $ keyczar addkey --location=/path/to/keyset --status=primary

3. Encrypt message with 'keyset':

  $ perl > msg.txt
  use Crypt::Keyczar::Crypter;
  my $kcz = Crypt::Keyczar::Crypter->new("/path/to/keyset");
  print $kcz->encrypt('Hello World!');
  __END__

SUPPORTED CRYPTOGRAPHIC ALGORITHMS(current version)
---------------------------------------------------
* HMAC
   Default keys are 256 bits. SHA1 used as the hash algorithm.
   HMAC-SHA224, HMAC-SHA256, HMAC-SHA384 and HMAC-SHA512
   algorithms are also supported.

* AES
   Default keys are 128 bits. 192 and 256 bit keys are also
   supported. CBC mode with PKCS#5 padding is used by default.

* RSA Encryption
   RSA-OAEP encryption is used. Default key size is 2048 bits.
   1024, 768, and 512 bit keys are also supported.

* RSA Signatures
   RSA-SHA1 signing is used. Default key size is 2048 bits.
   1024, 768, and 512 bit keys are also supported.

* DSA
    DSA-SHA1 signing algorithm used by default. Default key
    size is 1024 bits.

DEPENDENCIES
------------
This module requires these other modules and libraries:

  libcrypto (OpenSSL 0.9.8 or later)
  MIME::Base64
  JSON (version 1.xxx or 2.xxx)


SEE ALSO
--------
  "keyczar" in bin,
  perldoc Crypt::Keyczar,
  perldoc Crypt::Keyczar::Crypter,
  perldoc Crypt::Keyczar::Signer,
  http://www.keyczar.org/


Changes
0.07	Thu Oct 27 17:42:21 JST 2011
	- support perl 5.14.x and later version.
	- add java compat tests.
	- fixed Key lifecycle management bug, enable promote INACTIVE to ACTIVE.
	  https://rt.cpan.org/Public/Bug/Display.html?id=62032
	  Thanks to Freaking Wildchild http://wildchild.sesamid.com/

0.06	Wed Nov 18 17:47:22 JST 2009
	- fixed Revoke operation bug.
	  (There is not the compatibility of the keyset which "revoke" key)

0.05	Tue Nov 17 15:10:09 JST 2009
	- install 'bin/keyczar' tool. (thanks to Leon Brocard)
	- independent JSON::XS. (thanks to CPAN testers)

0.04	Thu Nov 12 19:16:26 JST 2009
	- suppress null signature warnings.
	- support expiration time.

0.03	Wed Nov 11 17:07:11 JST 2009
	- support HMAC-SHA224, HMAC-SHA256, HMAC-SHA384, HMAC-SHA512.

0.02	Tue Aug 18 17:04:34 JST 2009
	- rewrite Crypt::Keyczar::HmacEngine by XS.
	- rewrite Crypt::Keyczar::Util::random() by XS.
	- rewrite Crypt::Keyczar::AesEngine by XS.
	- support RSA private key decrypt.
	- support RSA public key encrypt.
	- support RSA private key verify.
	- support RSA private key sign.
	- support RSA public key verify.
	- support export RSA public key.
	- independent Digest::SHA1
	- add Crypt::Keyczar::FileWriter.
	- support DSA public key verify.
	- support DSA private key sign.

0.01	Wed Jul 22 20:15:10 JST 2009
	- add "usekey" option.
	- add Java and Python test code.
	- support "revoke".
	- support AES-CBC with HMAC-SHA1 encryption.
	- support HMAC-SHA1 sing/verify.
	- fixed create keyset and ad-key operation.


lib/Crypt/Keyczar/Key.pm
package Crypt::Keyczar::Key;
use strict;
use warnings;
use Carp;

use Crypt::Keyczar::Util;



my %POLICY = (
    'AES' => {
        object => 'Crypt::Keyczar::AesKey',
        keysize => [128, 192, 256],
        outsize => 0
    },
    'HMAC_SHA1' => {
        object => 'Crypt::Keyczar::HmacKey',
        keysize => [256],
        outsize => 20
    },
    'DSA_PRIV' => {
        object => 'Crypt::Keyczar::DsaPrivateKey',
        keysize => [1024],
        outsize => 48
    },
    'DSA_PUB' => {
        object => 'Crypt::Keyczar::DsaPublicKey',
        keysize => [1024],
        outsize => 48
    },
    'RSA_PRIV' => {
        object => 'Crypt::Keyczar::RsaPrivateKey',
        keysize => [2048, 1024, 768, 512],
        outsize => 256
    },
    'RSA_PUB' => {
        object => 'Crypt::Keyczar::RsaPublicKey',
        keysize => [2048, 1024, 768, 512],
        outsize => 256
    },

    'HMAC_SHA224' => {
        object => 'Crypt::Keyczar::HmacSHA224Key',
        keysize => [256],
        outsize => 28
    },
    'HMAC_SHA256' => {
        object => 'Crypt::Keyczar::HmacSHA256Key',
        keysize => [256],
        outsize => 32 
    },
    'HMAC_SHA384' => {
        object => 'Crypt::Keyczar::HmacSHA384Key',
        keysize => [256],
        outsize => 48
    },
    'HMAC_SHA512' => {
        object => 'Crypt::Keyczar::HmacSHA512Key',
        keysize => [256],
        outsize => 64
    },
);


sub set_policy {
    my $class = shift;
    %POLICY = @_;
}


sub new {
    my $class = shift;
    return bless {}, $class;
}


sub init {}


sub get_header {
    my $self = shift;
    return pack 'C1 a*', Crypt::Keyczar::FORMAT_VERSION(), $self->hash();
}

sub expose {
    my $self = shift;
    my $expose = {};
    $expose->{size} = $self->{size};
    return $expose;
}


sub hash {
    my $self = shift;
    $self->{__hash} = shift if @_;
    return $self->{__hash};
}


sub get_size {}
sub get_type {}
sub get_engine {}
sub get_sign_engine { return Crypt::Keyczar::_NullSignEngine->new(); }



sub generate_key {
    my $class = shift;
    my ($type, $size) = @_;

    if (!exists $POLICY{uc $type}) {
        croak "unsupported key type: $type";
    }

    my $key = $POLICY{uc $type}->{object};
    eval "use $key";
    if ($@) {
        croak "$type: $@";
    }
    if ($size) {
        if (scalar(grep {$_ == $size} @{$POLICY{uc $type}->{keysize}}) == 0) {
            croak "unacceptable key size: $size\@$type";
        }
    }
    else {
        $size = $POLICY{uc $type}->{keysize}->[0]; # set default key size
    }
    return $key->generate($size);
}


sub read_key {
    my $class = shift;
    my ($type, $json_string_key) = @_;

    if (!exists $POLICY{uc $type}) {
        croak "unsupported key type: $type";
    }
    my $key = $POLICY{uc $type}->{object};
    eval "use $key";
    if ($@) {
        croak "$type: $@";
    }
    return $key->read($json_string_key);
}


sub to_string {
    return Crypt::Keyczar::Util::encode_json($_[0]->expose);
}

1;


package Crypt::Keyczar::_NullSignEngine;
use strict;
use warnings;

sub new { return bless {}, $_[0]; }
sub digest_size { return 0 }
sub init { return '' }
sub update {}
sub sign { return '' }
sub verify { return 1 }



__END__
